data/txDupCache.go

```diff
@@ -191,51 +192,74 @@ func (c *txSaltedCache) innerCheck(msg []byte) (*crypto.Digest, bool) {
 
 	d := crypto.Digest(blake2b.Sum256(toBeHashed))
 
-	_, found := c.cur[d]
+	v, found := c.cur[d]
 	if found {
-		return nil, true
+		return &d, v, &c.cur, true
 	}
 
 	toBeHashed = append(toBeHashed[:len(msg)], c.prevSalt[:]...)
 	toBeHashed = toBeHashed[:len(msg)+len(c.prevSalt)]
 	pd := crypto.Digest(blake2b.Sum256(toBeHashed))
-	_, found = c.prev[pd]
+	v, found = c.prev[pd]
 	if found {
-		return nil, true
+		return &pd, v, &c.prev, true
 	}
-	return &d, false
+	return &d, nil, nil, false
 }
 
 // CheckAndPut adds msg into a cache if not found
 // returns a hashing key used for insertion if the message not found.
-func (c *txSaltedCache) CheckAndPut(msg []byte) (*crypto.Digest, bool) {
+func (c *txSaltedCache) CheckAndPut(msg []byte, sender network.Peer) (*crypto.Digest, *sync.Map, bool) {
 	c.mu.RLock()
-	d, found := c.innerCheck(msg)
+	d, vals, page, found := c.innerCheck(msg)
 	salt := c.curSalt
-	c.mu.RUnlock()
 	// fast read-only path: assuming most messages are duplicates, hash msg and check cache
+	// keep lock - it is needed for copying vals in defer
+	senderFound := false
 	if found {
-		return d, found
+		if _, senderFound = vals.Load(sender); senderFound {
+			c.mu.RUnlock()
+			return d, vals, true
+		}
 	}
+	c.mu.RUnlock()
 
 	// not found: acquire write lock to add this msg hash to cache
 	c.mu.Lock()
 	defer c.mu.Unlock()
 	// salt may have changed between RUnlock() and Lock(), rehash if needed
 	if salt != c.curSalt {
-		d, found = c.innerCheck(msg)
+		d, vals, page, found = c.innerCheck(msg)
 		if found {
-			// already added to cache between RUnlock() and Lock(), return
-			return d, found
+			if _, senderFound = vals.Load(sender); senderFound {
+				// already added to cache between RUnlock() and Lock(), return
+				return d, vals, true
+			}
 		}
-	} else {
+	} else if found && page == &c.prev {
+		// there is match with prev page, update the value with data possible added in between locks
+		vals, found = c.prev[*d]
+	} else { // not found or found in cur page
 		// Do another check to see if another copy of the transaction won the race to write it to the cache
-		// Only check current to save a lookup since swaps are rare and no need to re-hash
-		if _, found := c.cur[*d]; found {
-			return d, found
+		// Only check current to save a lookup since swap is handled in the first branch
+		vals, found = c.cur[*d]
+		if found {
+			if _, senderFound = vals.Load(sender); senderFound {
+				return d, vals, true
+			}
+			page = &c.cur
 		}
 	}
 
+	// at this point we know that either:
+	// 1. the message is not in the cache
+	// 2. the message is in the cache but from other senders
+	if found && !senderFound {
```

Comment:
According to the comment above, this should be sufficient:
```suggestion
	if found {
```